### PR TITLE
Fix ws halt

### LINKED ISSUE #6908: Worldstate download halt
Besu version 24.4.0-RC2 (seen three times in burnin nodes)

Including tests of main, this has been seen on 7 / 32 nodes = 22% occurrence 
 - this number is in flux as analysis still being done, as well as data being gathered.

### PR DESCRIPTION
**Pull request by jframe**
## PR description

PR created on behalf of @matkt https://github.com/matkt/besu/tree/feature/try-fix-ws-halt branch.

It fixes a possible halting of the worldstate download during the initial sync.

- Modified worldstate sync pipeline so it is notified when importing a new block. Before the pipeline got stuck in some cases
- Added a timeout of 10s on tasks
- Removed useless requests from retries. Old requests were accumulating before

### Testing
- Checkpoint sync on mainnet with 15 nodes. None had a worldstate halt with the PR. Before ~20% of nodes failed with ws halt.

## Fixed Issue(s)
Fixes #6908 

### Thanks for sending a pull request! Have you done the following?

- [ ] Checked out our [contribution guidelines](https://github.com/hyperledger/besu/blob/main/CONTRIBUTING.md)?
- [ ] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [ ] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [ ] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Locally, you can run these tests to catch failures early:

- [ ] unit tests: `./gradlew build`
- [ ] acceptance tests: `./gradlew acceptanceTest`
- [ ] integration tests: `./gradlew integrationTest`
- [ ] reference tests: `./gradlew ethereum:referenceTests:referenceTests`

### CODE REVIEW

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/CompleteTaskStep.java [hunk 1]
```diff
@@ -43,9 +43,11 @@ public CompleteTaskStep(
 
   public synchronized void markAsCompleteOrFailed(
       final SnapWorldDownloadState downloadState, final Task<SnapDataRequest> task) {
-    if (task.getData().isResponseReceived()
-        || (task.getData() instanceof TrieNodeHealingRequest
-            && task.getData().isExpired(snapSyncState))) {
+    final boolean isResponseReceived = task.getData().isResponseReceived();
+    final boolean isExpiredRequest =
+        task.getData() instanceof TrieNodeHealingRequest && task.getData().isExpired(snapSyncState);
+    // by the old one when we are changing pivot block
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/CompleteTaskStep.java:49]:
```suggestion
    // if pivot block has changed, the request is expired and we mark this one completed
```

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java [hunk 2]
```diff
@@ -111,7 +110,9 @@ public SnapWorldDownloadState(
     this.blockchain = blockchain;
     this.snapSyncState = snapSyncState;
     this.metricsManager = metricsManager;
-    this.blockObserverId = OptionalLong.empty();
+    this.blockObserverId = blockchain.observeBlockAdded(createBlockchainObserver());
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java:113]: nit - can the comment below go before this line

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java [hunk 3]
```diff
@@ -441,8 +437,7 @@ public BlockAddedObserver createBlockchainObserver() {
       final boolean isBlockchainCaughtUp =
           snapSyncState.isWaitingBlockchain() && !pivotBlockSelector.isBlockchainBehind();
 
-      if (isNewPivotBlockFound
-          || isBlockchainCaughtUp) { // restart heal if we found a new pivot block or if close to
+      if (snapSyncState.isHealTrieInProgress() && (isNewPivotBlockFound || isBlockchainCaughtUp)) {
         // head again
```
Comment (macfarla) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java:441]: the rest of this comment has been deleted - doesn't make sense by itself

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java (hunk 2 above)
Comment (macfarla) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java:113]: or delete - it's not really adding anything, the line of code says the same thing

**Review by macfarla — APPROVED** [review]
changes make sense. few nits on comments. could use a changelog entry.

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java (hunk 2 above)
Comment (jframe) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java:113]: Done. Removed it. Don't it was adding anything

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java (hunk 3 above)
Comment (jframe) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/SnapWorldDownloadState.java:441]: Done. Removed it, don't think we need the comment

**Pull request merged**